gc/scheduler/scheduler.go

```diff
@@ -0,0 +1,332 @@
+	RegisterMutationCallback(func(bool))
+	GarbageCollect(context.Context) (metadata.GCStats, error)
+}
+
+type gcScheduler struct {
+	c collector
+
+	eventC chan mutationEvent
+
+	waiterL sync.Mutex
+	waiters []chan metadata.GCStats
+
+	pauseThreshold    float64
+	deletionThreshold int
+	mutationThreshold int
+	scheduleDelay     time.Duration
+	startupDelay      time.Duration
+}
+
+func newScheduler(c collector, cfg *config) *gcScheduler {
+	eventC := make(chan mutationEvent)
+
+	s := &gcScheduler{
+		c:                 c,
+		eventC:            eventC,
+		pauseThreshold:    cfg.PauseThreshold,
+		deletionThreshold: cfg.DeletionThreshold,
+		mutationThreshold: cfg.MutationThreshold,
+		scheduleDelay:     time.Duration(cfg.ScheduleDelay),
+		startupDelay:      time.Duration(cfg.StartupDelay),
+	}
+
+	if s.pauseThreshold < 0.0 {
+		s.pauseThreshold = 0.0
+	}
+	if s.pauseThreshold > 0.5 {
+		s.pauseThreshold = 0.5
+	}
+	if s.mutationThreshold < 0 {
+		s.mutationThreshold = 0
+	}
+	if s.scheduleDelay < 0 {
+		s.scheduleDelay = 0
+	}
+	if s.startupDelay < 0 {
+		s.startupDelay = 0
+	}
+
+	c.RegisterMutationCallback(s.mutationCallback)
+
+	return s
+}
+
+func (s *gcScheduler) ScheduleAndWait(ctx context.Context) (metadata.GCStats, error) {
+	return s.wait(ctx, true)
+}
+
+func (s *gcScheduler) wait(ctx context.Context, trigger bool) (metadata.GCStats, error) {
+	wc := make(chan metadata.GCStats, 1)
+	s.waiterL.Lock()
+	s.waiters = append(s.waiters, wc)
+	s.waiterL.Unlock()
+
+	if trigger {
+		e := mutationEvent{
+			ts: time.Now(),
+		}
+		go func() {
+			s.eventC <- e
+		}()
+	}
+
+	var gcStats metadata.GCStats
+	select {
+	case stats, ok := <-wc:
+		if !ok {
+			return metadata.GCStats{}, errors.New("gc failed")
+		}
+		gcStats = stats
+	case <-ctx.Done():
+		return metadata.GCStats{}, ctx.Err()
+	}
+
+	return gcStats, nil
+}
+
+func (s *gcScheduler) mutationCallback(dirty bool) {
+	e := mutationEvent{
+		ts:       time.Now(),
+		mutation: true,
+		dirty:    dirty,
+	}
+	go func() {
+		s.eventC <- e
+	}()
+}
+
+func schedule(d time.Duration) (<-chan time.Time, *time.Time) {
+	next := time.Now().Add(d)
+	return time.After(d), &next
+}
+
+func (s *gcScheduler) run(ctx context.Context) {
+	var (
+		schedC <-chan time.Time
+
+		lastCollection *time.Time
+		nextCollection *time.Time
+
+		interval    = time.Second
+		gcTime      time.Duration
+		collections int
+		// TODO(dmcg): expose collection stats as metrics
+
+		triggered bool
+		deletions int
+		mutations int
+	)
+	if s.startupDelay > 0 {
+		schedC, nextCollection = schedule(s.startupDelay)
+	}
+	for {
+		select {
+		case <-schedC:
+			// Check if garbage collection can be skipped because
+			// it is not needed or was not requested and reschedule
+			// it to attempt again after another time interval.
+			if !triggered && lastCollection != nil && deletions == 0 &&
+				(s.mutationThreshold == 0 || mutations < s.mutationThreshold) {
+				schedC, nextCollection = schedule(interval)
```

Comment:
I know this is a _very_ old PR, but is it expected that this function is called with a 60ms `interval` on an idle `containerd`?

```
ivan@vm:~$ sudo strace -f -e futex -p $(pidof containerd) 2>&1 | fgrep tv_nsec
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=4609266}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=42788360}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58039972}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=59098241}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58143429}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58114054}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58206593}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=59888972}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58140470}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58126429}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=60019303}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58203135}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58190302}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58506879}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58253510} <unfinished ...>
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58054264}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58440797}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=59885181}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=60048094}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58289384}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=59167155}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58111304}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58379840}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58202719}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58246551}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58164469}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58169178}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58226094}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58159470}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58195511}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=60370420}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=58232218}) = -1 ETIMEDOUT (Connection timed out)
[pid 25650] futex(0x558299ab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=57952599}) = -1 ETIMEDOUT (Connection timed out)
```

This burns some CPU, 50ms over 10s on my M1 MacBook Air:

```
ivan@vm:~$ systemctl status containerd | grep CPU; sleep 10; systemctl status containerd | grep CPU
        CPU: 744ms
        CPU: 795ms
```

Here's a flamegraph:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/89186/223037172-733c2d48-d99f-4362-be0d-f18b8669c91b.png">

It disappears if I make `schedule()` sleep for an hour instead:

```
ivan@vm:~$ systemctl status containerd | grep CPU; sleep 10; systemctl status containerd | grep CPU
        CPU: 151ms
        CPU: 153ms
```

There's still other scheduled stuff in the flamegraph, but not as much:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/89186/223038247-a6de41d0-d4ab-4fb1-b0cf-debb45f3d073.png">

I think it's the 10s timer from here (not great, but also not as aggressive as gc here):

* https://github.com/containerd/containerd/blob/v1.7.0-rc.1/runtime/restart/monitor/monitor.go#L91

```
ivan@vm:~$ sudo strace -f -e futex -p $(pidof containerd) 2>&1 | fgrep tv_nsec
[pid 26061] futex(0x555ffcab90, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=100000} <unfinished ...>
[pid 26061] futex(0x555ffcab90, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=100000} <unfinished ...>
[pid 26054] futex(0x555ffcab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=9, tv_nsec=974986451}) = -1 ETIMEDOUT (Connection timed out)
[pid 26054] futex(0x555ffcab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=9, tv_nsec=982671012}) = -1 ETIMEDOUT (Connection timed out)
[pid 26054] futex(0x555ffcab78, FUTEX_WAIT_PRIVATE, 0, {tv_sec=9, tv_nsec=999289434}) = -1 ETIMEDOUT (Connection timed out)
```

I'm seeing the same on `v1.6.18` and the latest `HEAD`.

Reply:
It has been awhile and now we know a lot more about how long GC takes a to run. The GC usually runs much faster than we had originally anticipated, likely due to most deployments not using an unreasonable large number of images/layers.

The algorithm for computing the interval is still
```
			// Set interval to average gc time divided by the pause threshold
			// This algorithm ensures that a gc is scheduled to allow enough
			// runtime in between gc to reach the pause threshold.
			// Pause threshold is always 0.0 < threshold <= 0.5
			avg := float64(gcTime) / float64(collections)
			interval = time.Duration(avg/s.pauseThreshold - avg)
```

In your case, my guess is the average GC is be running in less than 2ms and it is using the default 0.02 threshold. My recommendation would be either lower the threshold even farther or just set it to 0.0 to keep the interval at 1 second. 

We could also explore updating this logic especially for idle systems. Ideally the scheduler could just sleep until there is a mutation or expiration (next expiration would need to be calculated during GC).